iroha/src/sumeragi.rs

```diff
@@ -1,14 +1,216 @@
-use crate::prelude::*;
+use crate::{
+    crypto::{Crypto, Hash},
+    peer::{self, PeerId},
+    prelude::*,
+};
+use iroha_derive::*;
+use parity_scale_codec::{Decode, Encode};
+use std::cmp::Ordering;
 
-#[derive(Default)]
-pub struct Sumeragi {}
+pub struct Sumeragi {
+    sorted_peers: Vec<PeerId>,
+    max_faults: usize,
+    peer_id: PeerId,
+    /// Block in discussion this round
+    pending_block: Option<Block>,
+}
+
+#[derive(Io, Decode, Encode, Debug, Clone)]
+pub enum Message {
+    /// Is sent by leader to all validating peers, when a new block is created.
+    Created(Block),
+    /// Is sent by validating peers to proxy tail and observing peers when they have signed this block.
+    Signed(Block),
+    /// Is sent by proxy tail to validating peers and to leader, when the block is committed.
+    Committed(Block),
+}
+
+#[derive(Eq, PartialEq, Debug)]
+pub enum Role {
+    Leader,
+    ValidatingPeer,
+    ObservingPeer,
+    ProxyTail,
+}
 
 impl Sumeragi {
-    pub fn new() -> Self {
-        Sumeragi {}
+    pub fn new(
+        peers: &[PeerId],
+        prev_block_hash: Option<Hash>,
+        max_faults: usize,
+    ) -> Result<Self, String> {
+        let min_peers = 3 * max_faults + 1;
+        if peers.len() >= min_peers {
+            let mut peers = peers.to_vec();
+            Self::sort_peers(&mut peers, prev_block_hash);
+            Ok(Self {
+                sorted_peers: peers,
+                max_faults,
+                //TODO: generate peer's public key, save on shutdown and load on start
+                peer_id: PeerId {
+                    address: "127.0.0.1:7878".to_string(),
+                    public_key: [0u8; 32],
+                },
+                pending_block: None,
+            })
+        } else {
+            Err(format!("Not enough peers to be Byzantine fault tolerant. Expected a least {} peers, got {}", 3 * max_faults + 1, peers.len()))
+        }
+    }
+
+    pub fn next_round(&mut self, prev_block_hash: Hash) {
+        Self::sort_peers(&mut self.sorted_peers, Some(prev_block_hash));
+        self.pending_block = None;
+    }
+
+    pub fn peers_set_a(&self) -> &[PeerId] {
+        let n_a_peers = 2 * self.max_faults + 1;
+        &self.sorted_peers[..n_a_peers]
     }
 
-    pub async fn sign(
+    pub fn peers_set_b(&self) -> &[PeerId] {
+        &self.sorted_peers[(2 * self.max_faults + 1)..]
+    }
+
+    pub fn leader(&self) -> &PeerId {
+        self.peers_set_a()
+            .first()
+            .expect("Failed to get first peer.")
+    }
+
+    pub fn proxy_tail(&self) -> &PeerId {
+        self.peers_set_a().last().expect("Failed to get last peer.")
+    }
+
+    pub fn validating_peers(&self) -> &[PeerId] {
+        &self.peers_set_a()[1..(self.peers_set_a().len() - 1)]
```

Comment:
what happens if the peers_set_a is Len 0?

Reply:
This can't happen as the `Sumeragi::new` function checks that the number of peers is more or equal to `3f+1`, so it's at least 1.